Flaky test: engine.processing.resource.ResourceDeletionTest.shouldWriteEventsForDeletedProcessWithBannedInstances

**Summary**

> Try to answer the following as best as possible

- How often does the test fail? once
- Does it block your work? no
- Do we suspect that it is a real failure? not sure

**Failures**

> Outline known failure cases, e.g. a failed assertion and its stacktrace obtained from Jenkins

<details><summary>Example assertion failure</summary>
<pre>
Warning:  Flakes: 
Warning:  io.camunda.zeebe.engine.processing.resource.ResourceDeletionTest.shouldWriteEventsForDeletedProcessWithBannedInstances
Error:    Run 1: ResourceDeletionTest.shouldWriteEventsForDeletedProcessWithBannedInstances:259 » Runtime Unexpected error occurred during RocksDB transaction.

---------

Error:  Tests run: 26, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 6.266 s <<< FAILURE! -- in io.camunda.zeebe.engine.processing.resource.ResourceDeletionTest
Error:  io.camunda.zeebe.engine.processing.resource.ResourceDeletionTest.shouldWriteEventsForDeletedProcessWithBannedInstances -- Time elapsed: 0.352 s <<< ERROR!
java.lang.RuntimeException: Unexpected error occurred during RocksDB transaction.
	at io.camunda.zeebe.db.impl.rocksdb.transaction.DefaultTransactionContext.runInTransaction(DefaultTransactionContext.java:42)
	at io.camunda.zeebe.db.impl.rocksdb.transaction.TransactionalColumnFamily.ensureInOpenTransaction(TransactionalColumnFamily.java:324)
	at io.camunda.zeebe.db.impl.rocksdb.transaction.TransactionalColumnFamily.upsert(TransactionalColumnFamily.java:111)
	at io.camunda.zeebe.engine.state.processing.DbBannedInstanceState.banInstance(DbBannedInstanceState.java:68)
	at io.camunda.zeebe.engine.state.processing.DbBannedInstanceState.banProcessInstance(DbBannedInstanceState.java:114)
	at io.camunda.zeebe.engine.processing.resource.ResourceDeletionTest.shouldWriteEventsForDeletedProcessWithBannedInstances(ResourceDeletionTest.java:259)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
Caused by: org.rocksdb.RocksDBException: WriteBatch has wrong count
	at org.rocksdb.Transaction.commit(Native Method)
	at org.rocksdb.Transaction.commit(Transaction.java:207)
	at io.camunda.zeebe.db.impl.rocksdb.transaction.ZeebeTransaction.commitInternal(ZeebeTransaction.java:136)
	at io.camunda.zeebe.db.impl.rocksdb.transaction.DefaultTransactionContext.runInNewTransaction(DefaultTransactionContext.java:62)
	at io.camunda.zeebe.db.impl.rocksdb.transaction.DefaultTransactionContext.runInTransaction(DefaultTransactionContext.java:33)
	... 6 more
</pre>
</details>

**Hypotheses**

> List any hypotheses if you have one; can be ommitted

**Logs**

> If possible, provide more context here, e.g. standard output logs, link to build, etc.

Build link: https://github.com/camunda/zeebe/actions/runs/7019579999/job/19097314743?pr=15365

--- a/zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/resource/ResourceDeletionTest.java
+++ b/zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/resource/ResourceDeletionTest.java
@@ -15,6 +15,7 @@ import static org.assertj.core.api.Assertions.tuple;
 import io.camunda.zeebe.engine.state.mutable.MutableBannedInstanceState;
 import io.camunda.zeebe.engine.util.EngineRule;
 import io.camunda.zeebe.model.bpmn.Bpmn;
+import io.camunda.zeebe.protocol.Protocol;
 import io.camunda.zeebe.protocol.record.Record;
 import io.camunda.zeebe.protocol.record.intent.DecisionEvaluationIntent;
 import io.camunda.zeebe.protocol.record.intent.DecisionIntent;
@@ -251,12 +252,17 @@ public class ResourceDeletionTest {
             .get(0)
             .getProcessDefinitionKey();
     final long processInstanceKey = engine.processInstance().ofBpmnProcessId(processId).create();
+    final var partitionId = Protocol.decodePartitionId(processInstanceKey);
 
     // Note! We don't register the banned instance using an event. You won't see the Error Event in
     // the log!
+    // We need to pause processing to prevent potential transaction corruption when modifying the
+    // state from the test thread concurrently to processing.
+    engine.pauseProcessing(partitionId);
     final var bannedInstanceState =
         (MutableBannedInstanceState) engine.getProcessingState().getBannedInstanceState();
     bannedInstanceState.banProcessInstance(processInstanceKey);
+    engine.resumeProcessing(partitionId);
 
     // when
     engine.resourceDeletion().withResourceKey(processDefinitionKey).delete();

